cleverhans/experimental/certification/dual_formulation.py

```diff
@@ -344,14 +457,33 @@ def np_vector_prod_fn_h(np_vector):
                                np.multiply(self.switch_indices[i],
                                            np.maximum(new_lambda_neg_val[i], 0)))
 
-    dual_feed_dict.update(zip(self.lambda_lu, new_lambda_lu_val))
-    dual_feed_dict.update(zip(self.lambda_neg, new_lambda_neg_val))
-    scalar_f = self.sess.run(self.scalar_f, feed_dict=dual_feed_dict)
-    vector_g = self.sess.run(self.vector_g, feed_dict=dual_feed_dict)
-    x, _ = lgmres(linear_operator_h, vector_g)
-    x = x.reshape((x.shape[0], 1))
-    second_term = np.matmul(np.transpose(vector_g), x) + 0.05
+    # Assign new lambda
+    # TODO(shankarshreya): remove the assign ops and do this with new variables
+    self.sess.run([tf.assign(var, val) for var, val in zip(self.lambda_lu, new_lambda_lu_val)])
+    self.sess.run([tf.assign(var, val) for var, val in zip(self.lambda_neg, new_lambda_neg_val)])
+
+    # Make matrix M PSD
+    scalar_f = self.sess.run(self.scalar_f)
+    old_nu, second_term = self.make_m_psd()
 
     computed_certificate = scalar_f + 0.5*second_term
 
-    return computed_certificate
+    tf.logging.info('Inner step: %d, current value of certificate: %f',
+                    current_step, computed_certificate)
+
+    # Sometimes due to either overflow or instability in inverses,
+    # the returned certificate is large and negative -- keeping a check
+    if LOWER_CERT_BOUND < computed_certificate < -1:
+      _, min_eig_val_m = self.get_lanczos_eig()
+      if min_eig_val_m - TOL > 0:
+        tf.logging.info('Found certificate of robustness!')
+        return True
+
+    # Reset values of the variables to their original values before projection
+    # If we don't do this, the optimizer will use the projected values for variables
+    # instead of the true values
+    self.sess.run([tf.assign(var, val) for var, val in zip(self.lambda_lu, lambda_lu_val)])
+    self.sess.run([tf.assign(var, val) for var, val in zip(self.lambda_neg, lambda_neg_val)])
+    self.sess.run(tf.assign(self.nu, old_nu))
```

Comment:
As we discussed offline, it's better to recover old value of nu inside `make_m_psd` so this function won't have side effects